module-bsp/board/rt1051/bellpx/hal/battery_charger/BatteryCharger.cpp

```diff
@@ -1,86 +1,217 @@
     {
-        bsp::fuel_gauge::init(queueBatteryHandle);
-        bsp::battery_charger::init(queueBatteryHandle);
+      public:
+        enum class IrqEvents : std::uint8_t
+        {
+            Charger,
+            FuelGauge
+        };
+        using BatteryWorkerQueue = WorkerQueue<IrqEvents>;
+
+        explicit BellBatteryCharger(xQueueHandle irqQueueHandle);
+        ~BellBatteryCharger();
+
+        Voltage getBatteryVoltage() const final;
+        SOC getSOC() const final;
+        ChargingStatus getChargingStatus() const final;
+
+        static BatteryWorkerQueue &getWorkerQueueHandle();
+
+      private:
+        void sendNotification(Events event);
+        SOC fetchBatterySOC() const;
+        void pollFuelGauge();
+
+        xTimerHandle emergency_poll_timer;
+        xQueueHandle notification_channel;
+        inline static std::unique_ptr<BatteryWorkerQueue> worker_queue;
+
+        std::shared_ptr<drivers::DriverGPIO> charger_gpio_chgok;
+        std::shared_ptr<drivers::DriverGPIO> charger_gpio;
+        mutable bsp::devices::power::MP2639B charger;
+
+        std::shared_ptr<drivers::DriverGPIO> fuel_gauge_gpio;
+        mutable bsp::devices::power::CW2015 fuel_gauge;
+    };
+
+    BellBatteryCharger::BellBatteryCharger(xQueueHandle irqQueueHandle)
+        : notification_channel{irqQueueHandle}, charger_gpio_chgok{drivers::DriverGPIO::Create(charger_irq_gpio_chgok,
+                                                                                               {})},
+          charger_gpio{drivers::DriverGPIO::Create(charger_irq_gpio, {})}, charger{MP2639B::Configuration{
+                                                                               charger_gpio,
+                                                                               charger_irq_pin_mode,
+                                                                               charger_gpio,
+                                                                               charger_irq_pin_acok,
+                                                                               charger_gpio_chgok,
+                                                                               charger_irq_pin_chgok,
+                                                                               [this](const auto) {
+                                                                                   sendNotification(Events::Charger);
+                                                                               }}},
+
+          fuel_gauge_gpio{drivers::DriverGPIO::Create(fuel_gauge_irq_gpio_instance, {})},
+          fuel_gauge{CW2015{drivers::DriverI2C::Create(i2c_instance, i2c_params)}}
+    {
+
+        emergency_poll_timer = xTimerCreate("emg_timer", emergency_poll_time, pdFALSE, this, [](TimerHandle_t xTimer) {
+            BellBatteryCharger *inst = static_cast<BellBatteryCharger *>(pvTimerGetTimerID(xTimer));
+            inst->pollFuelGauge();
+        });
```

Comment:
just stating: each created timer with freertos API can hang our general system timers in worst case and lessens response time for timers :) this is because system timers use freertos timer queue to notify Services about timer event (and then service executes the timer function connected )

Reply:
That's a really good point. It would be a much better solution(as you pointed out) to not use freertos timers directly but to wrap them with a container and send only a notification from the actual timer handler to some kind of queue(we can of course reuse or slightly modify the existing `WorkerQueue` or just create completely new mechanism).